h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostGPUCVModelBuilder.java

```diff
@@ -12,35 +12,77 @@ public class XGBoostGPUCVModelBuilder extends CVModelBuilder {
 
     private static final Logger LOG = Logger.getLogger(XGBoostGPUCVModelBuilder.class);
 
-    private final Deque<Integer> availableGpus;
-
+    private final GPUAllocator _allocator;
+    
     public XGBoostGPUCVModelBuilder(
-        Job job,
+        Job<?> job,
         ModelBuilder<?, ?, ?>[] modelBuilders,
         int parallelization,
         int[] gpuIds
     ) {
         super(job, modelBuilders, parallelization);
+        final List<Integer> availableGpus;
         if (gpuIds != null && gpuIds.length > 0) {
             availableGpus = new LinkedList<>();
             for (int id : gpuIds) availableGpus.add(id);
         } else {
             availableGpus = new LinkedList<>(GpuUtils.allGPUs());
         }
-        LOG.info("Using parallel GPU building on " + availableGpus.size() + " GPUs.");
+        LOG.info("Available #GPUs for CV model training: " + availableGpus.size());
+        _allocator = new GPUAllocator(availableGpus);
     }
 
     @Override
     protected void prepare(ModelBuilder<?, ?, ?> m) {
         XGBoost xgb = (XGBoost) m;
-        xgb._parms._gpu_id = new int[] { availableGpus.pop() };
+        xgb._parms._gpu_id = new int[] { _allocator.takeLeastUtilizedGPU() };
         LOG.info("Building " + xgb.dest() + " on GPU " + xgb._parms._gpu_id[0]);
     }
 
     @Override
     protected void finished(ModelBuilder<?, ?, ?> m) {
         XGBoost xgb = (XGBoost) m;
-        availableGpus.push(xgb._parms._gpu_id[0]);
+        _allocator.returnGPU(xgb._parms._gpu_id[0]);
+    }
+
+    static class GPUAllocator {
+        final int[] _gpu_utilization;
+
+        GPUAllocator(List<Integer> gpuIds) {
+            this(initUtilization(gpuIds));
+        }
+
+        GPUAllocator(int[] gpuUtilization) {
+            _gpu_utilization = gpuUtilization;
+        }
+        
+        static int[] initUtilization(List<Integer> gpus) {
+            final int maxGpuId = gpus.stream().max(Integer::compareTo)
+                    .orElseThrow(() -> new IllegalStateException("There are no GPUs available for XGBoost (" + gpus + ")."));
+            final int[] utilization = new int[maxGpuId + 1];
+            Arrays.fill(utilization, -1);
+            gpus.forEach(id -> utilization[id] = 0);
+            return utilization;
+        }
+
+        void returnGPU(int id) {
```

Comment:
Not sure whether _return_(GPU) is the best antonym to _take_(LeastUtilizedGPU). IMHO, it does not remind you the internally decreased counter on the first sight. I would find _release_(GPU) more convenient naming in the context of resource usage.

Reply:
I think you are right, fixed. Thanks for the suggestion.